### PR TITLE
Update grpo_dense_v2_flash.py

### PR DESCRIPTION
**Pull request by ChinmayK0607**
Added RoPE replacing absolute PE and QK norm for stable and efficient training avoiding too large of qk values

### CODE REVIEW

File: grpo_dense_v2_flash.py [hunk 1]
```diff
@@ -49,218 +47,212 @@
 # Mixed precision
 USE_AMP = True                    
 AMP_DTYPE = "bfloat16"           
-# USE_FLASH_ATTN is set by the auto-configuration above            
+# USE_FLASH_ATTN is set by the auto-configuration above      
+
+
+
```
Comment (tokenbender) [inline grpo_dense_v2_flash.py:53]: keep original, unnecessary extra line

File: grpo_dense_v2_flash.py [hunk 2]
```diff
@@ -49,218 +47,212 @@
...
+import wandb
+import modal
 
-stub = modal.App("grpo-dense-v2-reference-logging")
+stub = modal.App("grpo-dense-v2-rope-qknorm")
 
-# Modal image with Flash Attention
-flash_attn_wheel = "https://github.com/Dao-AILab/flash-attention/releases/download/v2.6.3/flash_attn-2.6.3+cu123torch2.3cxx11abiFALSE-cp311-cp311-linux_x86_64.whl"
+# Flash‑Attention wheel (CUDA 12.3 + Torch 2.3). Change if needed.
+FLASH_WHEEL = (
```
Comment (tokenbender) [inline grpo_dense_v2_flash.py:73]: all these non-functional changes should be removed like indentation mismatches

File: grpo_dense_v2_flash.py [hunk 3]
```diff
@@ -49,218 +47,212 @@
...
     def next(self):
-        # Ensure we have enough context
-        if self.pos + self.B*self.T + 1 > len(self.data):
+        if self.pos + self.B * self.T + 1 > len(self.data):
             self.pos = self.context_len
-            
-        contexts = []
-        targets = []
-        
+        ctx, tgt = [], []
```
Comment (tokenbender) [inline grpo_dense_v2_flash.py:131]: keep context and target as variables. messes up the var usage.

File: grpo_dense_v2_flash.py [hunk 4]
```diff
@@ -49,218 +47,212 @@
...
-            q, k, v = qkv.unbind(dim=2)
-            q = q.transpose(1, 2)  # (B, n_head, T, head_dim)
-            k = k.transpose(1, 2)
-            v = v.transpose(1, 2)
-            
-            # Scaled dot-product attention
-            scale = self.head_dim ** -0.5
-            scores = torch.matmul(q, k.transpose(-2, -1)) * scale
-            
+            scores = torch.matmul(q, k.transpose(-2, -1))  # (B, h, T, T)
```
Comment (tokenbender) [inline grpo_dense_v2_flash.py:187]: if flash attention is not present, the else part should use memory efficient standard attention as present in existing script

File: grpo_dense_v2_flash.py [hunk 5]
```diff
@@ -49,218 +47,212 @@
...
+            except Exception:
                 pass
-                
-    def forward(self, x, mask=None):
+
+    @staticmethod
+    def _qk_norm(q: torch.Tensor, k: torch.Tensor, eps: float = 1e-6):
+        q = q * torch.rsqrt(q.pow(2).mean(dim=-1, keepdim=True) + eps)
+        k = k * torch.rsqrt(k.pow(2).mean(dim=-1, keepdim=True) + eps)
+        return q, k
```
Comment (tokenbender) [inline grpo_dense_v2_flash.py:167]: if we use pytorch built in here, we can get efficient and numerically stable implementation

if we use L2 norm
`def _qk_norm(q: torch.Tensor, k: torch.Tensor):
    return F.normalize(q, dim=-1), F.normalize(k, dim=-1)`
    
if we use rms norm
`def _qk_norm(q: torch.Tensor, k: torch.Tensor):
    return F.rms_norm(q, (q.size(-1),)), F.rms_norm(k, (k.size(-1),))`

File: grpo_dense_v2_flash.py [hunk 6]
```diff
@@ -49,218 +47,212 @@
...
-        return logits
+        x = self.drop(self.wte(idx))
+        m = self.mask[:T, :T]
+        for blk in self.blocks:
+            x = blk(x, m)
+        x = self._rms(x)
+        return self.head(x)
+
+
+\
```
Comment (tokenbender) [inline grpo_dense_v2_flash.py:255]: stray backslash

File: grpo_dense_v2_flash.py (hunk 4 above)
Comment (ChinmayK0607) [reply grpo_dense_v2_flash.py:187]: Using F.scaled_dot_product  attention, uses an internal optimised kernal instead of a matmul

**Comment by ChinmayK0607** [comment]
resolved all comments, upgraded torch version for inbuilt rms norm.
